### PR TITLE
79 replace links that use http

### PR DESCRIPTION
**Pull request by ahmad-shahid**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -13,7 +13,10 @@
   },
   "author": "ESDC-EDSC",
   "license": "MIT",
-  "dependencies": {},
+  "dependencies": {
+    "axios": "^0.21.4",
```
Comment (nschonni) [inline package.json:17]: I think these are still devDependencies

File: TestLinks.js [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+                //Issue error message if URL starts with HTTP unless it is part of the HTTP exception list
+                if (url[i].startsWith("http:/") && !exceptionHTTPList.some(exceptionHTTPList => url[i].startsWith(exceptionHTTPList))){
+                    console.error("HTTP URL Found: " + url[i]);
+                    errorCount++;
+                }
+                else {
+                    try{
+                        validURL = new URL(url[i]);
+                        //We can choose to skip testing for intranet links
+                        if (!(process.env.DISABLE_PROXY && ((validURL.host).split(".")[1] == "prv" || intranetLinksList.some(intranetLinksList => validURL.href.startsWith(intranetLinksList))))){
```
Comment (plup2) [inline TestLinks.js:96]: Change .split with .endsWith('.prv')

File: TestLinks.js [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        files = files.concat(getFiles(directory));
+    });
+
+    console.log("***** Reading files and scanning for links");
+    for (let i = 0; i < files.length; i++) {
+        let fileMatch = fs.readFileSync(files[i], 'utf8').match(regex);
+        if (fileMatch != null) matches = matches.concat(fileMatch);
+    }
+
+    let url = [...new Set(matches)];
```
Comment (plup2) [inline TestLinks.js:127]: Rename `url` to `urls`

File: TestLinks.js [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+    console.log("***** Validating links");
+    validateLinks().then((errorCount) => {if (errorCount !=0){ 
+		throw new Error(errorCount + " error(s) were found when validating URLs.");
```
Comment (plup2) [inline TestLinks.js:131]: Include total number of links checked in error/exception message.

File: TestLinks.js [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+		done(false);
+	} else {
+		done();
+	}});
+}
```
Comment (plup2) [inline TestLinks.js:136]: At the end, add a console.log message sayingsomething like "Done, xxx URLs checked successfully"

File: TestLinks.js [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+            if (fs.statSync(name).isDirectory()){
+                getFiles(name, myFiles);
+            } else {
+                myFiles.push(name);
+            }
+        }
+        return myFiles;
+    }
+
+    async function validateLinks(){
```
Comment (plup2) [inline TestLinks.js:82]: `validateLinks` should take `urls` as a parameter

**Review by plup2 — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (plup2) [reply package.json:17]: (I already had started a review and commented on that change, so I resolved this one)

File: src/gcintranet/wet-header.js [hunk 7]
```diff
@@ -18,7 +18,7 @@ wet.builder.environmentHost = function(cdnEnv) {
         return '/';
     }
     else if (cdnEnv != null && cdnEnv.startsWith('http')) {
-        return cdnEnv; //if starts with http: return as-is
+        return cdnEnv; //if starts with https: return as-is
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply src/gcintranet/wet-header.js]: done

File: package.json [hunk 8]
```diff
@@ -13,7 +13,10 @@
   },
   "author": "ESDC-EDSC",
   "license": "MIT",
-  "dependencies": {},
+  "dependencies": {
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply package.json]: done

File: Gruntfile.js [hunk 9]
```diff
@@ -100,6 +101,13 @@ module.exports = function(grunt) {
         return true;
     });
 
+    grunt.registerTask('test-links', 'Test all links in files in the src and public (minus the WET folder) directories', function(target) {
+        if (!target || target === 'test') {
+            var done = this.async();
+            testFileLinks(done);
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply Gruntfile.js]: done

File: TestLinks.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply TestLinks.js:137]: done

File: TestLinks.js [hunk 10]
```diff
@@ -0,0 +1,136 @@
+module.exports = function testFileLinks(done)
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply TestLinks.js]: done

File: TestLinks.js [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    console.log("***** Reading files and scanning for links");
+    for (let i = 0; i < files.length; i++) {
+        let fileMatch = fs.readFileSync(files[i], 'utf8').match(regex);
+        if (fileMatch != null) matches = matches.concat(fileMatch);
+    }
+
+    let url = [...new Set(matches)];
+
+    console.log("***** Validating links");
+    validateLinks().then((errorCount) => {if (errorCount !=0){
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply TestLinks.js]: done

File: TestLinks.js [hunk 12]
```diff
@@ -0,0 +1,136 @@
+module.exports = function testFileLinks(done)
+{
+    process.env['NODE_TLS_REJECT_UNAUTHORIZED'] = 0;
+
+    const fs = require('fs');
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply TestLinks.js]: done

File: TestLinks.js [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+    const regex = /http[s]?:\/\/.*?(?="|')/g;
+    const axios = require('axios');
+    const ProxyAgent = require('https-proxy-agent');
+    const agent = ProxyAgent('http://proxy.prv:80');
+
+    let validURL;	
+    let directories = ["./src/", "./public/common/", "./public/gcintranet/", "./public/gcweb/", "./public/global/"];
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply TestLinks.js]: done

File: TestLinks.js [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+    let matches = [];
+    let config = (process.env.DISABLE_PROXY) ? null : {httpsAgent: agent, proxy: false};
+
+    //Exception list (complete skip of validation)
+    //Includes links found on legacy templates, links that require credentials and partial URLs
+    let exceptionSyntaxList = ["https://www.canada.ca/etc/designs/canada/cdts/gcweb/${definition.themeVersion}",
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply TestLinks.js]: done

File: TestLinks.js [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+                            "https://www.gcpedia.gc.ca",
+                            "https://gcdirectory",
+                            "https://templates.service",
+                            "https://gcannuaire-gcdirectory.gc.ca",
+                            "https://kmt-ogc.service.gc.ca",
+                            "http://agora.on",
+                            "https://gcintranet.tpsgc-pwgsc.gc.ca",
+                            "https://nscc-cnas.pwgsc-tpsgc.gc.ca",
+                            "https://portal-portail.tbs-sct.gc.ca",
+                            "https://nscc-cnas.pwgsc-tpsgc.gc.ca"]
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply TestLinks.js:55]: done

File: TestLinks.js [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+
+    //Get all files from a directory
+    function getFiles (dir, myFiles){
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply TestLinks.js]: done

File: TestLinks.js [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+    const fs = require('fs');
+    const regex = /http[s]?:\/\/.*?(?="|')/g;
+    const axios = require('axios');
+    const ProxyAgent = require('https-proxy-agent');
+    const agent = ProxyAgent('http://proxy.prv:80');
+
+    let validURL;	
+    let directories = ["./src/", "./public/common/", "./public/gcintranet/", "./public/gcweb/", "./public/global/"];
+    let matches = [];
+    let config = (process.env.DISABLE_PROXY) ? null : {httpsAgent: agent, proxy: false};
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply TestLinks.js]: done

File: TestLinks.js [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+            }
+        }
+        return myFiles;
+    }
+
+    async function validateLinks(){
+        let errorCount = 0;
+        for (let i =0; i<url.length; i++){
+            //Check if URL is in the exception list of syntax check and validation
+            if (!exceptionSyntaxList.some(exceptionSyntaxList => url[i].startsWith(exceptionSyntaxList))){
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply TestLinks.js]: done

File: TestLinks.js [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    async function validateLinks(){
+        let errorCount = 0;
+        for (let i =0; i<url.length; i++){
+            //Check if URL is in the exception list of syntax check and validation
+            if (!exceptionSyntaxList.some(exceptionSyntaxList => url[i].startsWith(exceptionSyntaxList))){	
+                //Issue error message if URL starts with HTTP unless it is part of the HTTP exception list
+                if (url[i].startsWith("http:/") && !exceptionHTTPList.some(exceptionHTTPList => url[i].startsWith(exceptionHTTPList))){
+                    console.error("HTTP URL Found: " + url[i]);
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply TestLinks.js]: done for all

File: TestLinks.js [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+                            try {
+                                const res = await axios.get(validURL.href, config);
+                                if (/4\d[^\D1]/.test(res.status) === true) console.log(validURL.href + "does not have a valid response. Status: " + res.status);
+                            } catch(err) { 
+                                console.error(validURL.href + " encountered the following error: " + err.message);
+                                errorCount++;
+                            }
+                        }
+                    } catch (err) {
+                        console.error(err);
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply TestLinks.js]: done

File: TestLinks.js (hunk 2 above)
Comment (ahmad-shahid) [reply TestLinks.js]: done

File: TestLinks.js [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+                    errorCount++;
+                }
+                else {
+                    try{
+                        validURL = new URL(url[i]);
+                        //We can choose to skip testing for intranet links
+                        if (!(process.env.DISABLE_PROXY && ((validURL.host).split(".")[1] == "prv" || intranetLinksList.some(intranetLinksList => validURL.href.startsWith(intranetLinksList))))){
+                            try {
+                                const res = await axios.get(validURL.href, config);
+                                if (/4\d[^\D1]/.test(res.status) === true) console.log(validURL.href + "does not have a valid response. Status: " + res.status);
```
[the remark this answers is not in the document]
Comment (ahmad-shahid) [reply TestLinks.js]: done

File: TestLinks.js (hunk 3 above)
Comment (ahmad-shahid) [reply TestLinks.js]: done

File: TestLinks.js (hunk 4 above)
Comment (ahmad-shahid) [reply TestLinks.js]: done

File: TestLinks.js (hunk 5 above)
Comment (ahmad-shahid) [reply TestLinks.js:137]: done

File: TestLinks.js (hunk 6 above)
Comment (ahmad-shahid) [reply TestLinks.js]: done

File: TestLinks.js [hunk 22]
```diff
@@ -0,0 +1,137 @@
...
+                                const res = await axios.get(validURL.href, config);
+                                //Check response - regex determines if response starts with '2' or '3' followed by 2 digits, or is 401 (this will be considered valid)
+                                if (/([23][\d]{2})|401/.test(res.status) === false) console.log("Error: " + validURL.href + "does not have a valid response. Status: " + res.status);
+                            } catch(err) { 
+                                console.error("Error: " + validURL.href + " encountered the following error: " + err.message);
+                                errorCount++;
+                            }
+                        }
+                    } catch (err) {
+                        console.error("Error: An error occured with URL: " + urls[i] + err);
```
Comment (plup2) [inline TestLinks.js:107]: In this message, there will be a missing space ` ` or `: ` delimiter between the url and the error message.

File: TestLinks.js [hunk 23]
```diff
@@ -0,0 +1,137 @@
...
+    for (let i = 0; i < files.length; i++) {
+        let fileMatches = fs.readFileSync(files[i], 'utf8').match(regex);
+        if (fileMatches != null) matches = matches.concat(fileMatches);
+    }
+
+    let urls = [...new Set(matches)];
+
+    console.log("***** Validating links");
+    const errorCount = await validateLinks(urls);
+	if (errorCount !=0) {
```
Comment (plup2) [inline TestLinks.js:132]: (would not comment on this if it was the only comment but while we're here): Indentation :)  (same for line 6)

File: TestLinks.js (hunk 22 above)
Comment (ahmad-shahid) [reply TestLinks.js]: Added a space

File: TestLinks.js (hunk 23 above)
Comment (ahmad-shahid) [reply TestLinks.js]: Fixed

**Review by plup2 — APPROVED** [review]

**Comment by plup2** [comment]
@ahmad-shahid  All changes are OK but PR is blocked because not all the commits were signed...

**Pull request merged**